source/slang/slang.cpp

```diff
@@ -895,35 +895,65 @@ SLANG_NO_THROW slang::TypeLayoutReflection* SLANG_MCALL Linkage::getTypeLayout(
     return asExternal(typeLayout);
 }
 
-SLANG_NO_THROW slang::TypeLayoutReflection* SLANG_MCALL Linkage::getParameterBlockLayout(
+SLANG_NO_THROW slang::TypeReflection* SLANG_MCALL Linkage::getContainerType(
     slang::TypeReflection* inType,
-    SlangInt targetIndex,
-    slang::LayoutRules rules,
+    slang::ContainerType containerType,
     ISlangBlob** outDiagnostics)
 {
     auto type = asInternal(inType);
 
-    if (targetIndex < 0 || targetIndex >= targets.getCount())
-        return nullptr;
-
-    auto target = targets[targetIndex];
-
-    // TODO: We need a way to pass through the layout rules
-    // that the user requested (e.g., constant buffers vs.
-    // structured buffer rules). Right now the API only
-    // exposes a single case, so this isn't a big deal.
-    //
-    SLANG_UNUSED(rules);
-
-    auto typeLayout = target->getParameterBlockLayout(type);
+    Type* containerTypeReflection = nullptr;
+    ContainerTypeKey key = {inType, containerType};
+    if (!m_containerTypes.TryGetValue(key, containerTypeReflection))
+    {
+        switch (containerType)
+        {
+        case slang::ContainerType::ConstantBuffer:
+            {
+                ConstantBufferType* cbType = getASTBuilder()->create<ConstantBufferType>();
+                cbType->elementType = type;
+                containerTypeReflection = cbType;
+            }
+            break;
+        case slang::ContainerType::ParameterBlock:
+            {
+                ParameterBlockType* pbType = getASTBuilder()->create<ParameterBlockType>();
+                pbType->elementType = type;
+                containerTypeReflection = pbType;
+            }
+            break;
+        case slang::ContainerType::StructuredBuffer:
+            {
+                HLSLStructuredBufferType* sbType =
+                    getASTBuilder()->create<HLSLStructuredBufferType>();
+                sbType->elementType = type;
+                containerTypeReflection = sbType;
+            }
+            break;
```

Comment:
I'm concerned about how this kind of thing works in practice. All the types here are derived from `DeclRefType`, but nothing here is setting up a valid `DeclRef`. That means that the types that come out of this function will be missing key pieces (including generic specialization arguments) that would be present on equivalent types coming from the parsing and semantic checking logic of the full compiler.

If you look at a similar operation like `ASTBuilder::getVectorType`, you'll find that it goes to some lengths to ensure that when it constructs a type for `vector<T,N>` it does so by constructing a full decl-ref to the declaration of `Vector` in the Slang stdlib, with proper generic arguments.

Obviously if this approach works for now, we can run with it.

Reply:
Right, currently no where in the compiler is using that decl ref so it runs correctly. If we run into any issues we should make sure to fix this.